### PR TITLE
[python] Bump Python minimum version to 3.8

### PR DESCRIPTION
**Pull request by linux4life798**
As per the RFC https://discourse.llvm.org/t/rfc-upgrading-llvms-minimum-required-python-version/67571,
raise the minimum Python version to ensure that the Python syntax doesn't
become overly obsolete, to enable new Python feature usage, and to improve
the maintainability of CI.

One of the primary use cases for this higher Python version is to enable python type
annotations that are more aligned with current Python best practices. This is not only
important for our own internal Python for testing, but for the Python bindings that
are exposed to users.

### CODE REVIEW

File: llvm/docs/GettingStarted.rst
Comment (boomanaiden154) [inline llvm/docs/GettingStarted.rst:297]: I would change up the wording here. There are many other facilities within LLVM that also need Python. I'd specifically mention the test suite and then wrap everything else into a generic term.

**Review by boomanaiden154 — COMMENTED** [review]
There are a couple other files that should probably be updated:
* `llvm/docs/GettingStartedVS.rst`
* `llvm/docs/TestingGuide.rst`

That looks like it should be it when I searched for `3.6` within the repository.

File: llvm/docs/GettingStarted.rst
Comment (linux4life798) [reply llvm/docs/GettingStarted.rst]: Checkout the latest commits. I gave it a shot, but I might not have enough context of other LLVM Python projects to accurately categorize them.

Comment (linux4life798) [reply llvm/docs/GettingStarted.rst]: Please reopen if you think this can be further improved.

Comment (boomanaiden154) [reply llvm/docs/GettingStarted.rst]: LGTM. I think that should encompass the uses in tree, even including subprojects like LLDB, although I guess those bindings aren't fully auxiliary as they're on by default. The LLDB build documentation is separate though.

Comment (linux4life798) [reply llvm/docs/GettingStarted.rst]: True. I'm more than happy to change the sentence to correctly represent the state of the world.

Feel free to suggest or maybe the following?

```
Only needed if you want to run the automated test suite in the
``llvm/test`` directory, or if you plan to utilize any Python libraries or utilities.
```

or "LLVM Python libraries or utilities."

**Comment by tru** [comment]
This need a release note as well!

**Review by jpienaar — COMMENTED** [review]
(Biasedly) looks good to me with release notes added.

File: llvm/docs/GettingStarted.rst
Comment (boomanaiden154) [reply llvm/docs/GettingStarted.rst]: I'd mention bindings in their too and then I think that should pretty much cover everything:
```
...or if you plan to utilize any Python libraries, utilities, or bindings.
```

**Comment by tru** [comment]
I think it needs to go here: https://llvm.org/docs/ReleaseNotes.html#update-on-required-toolchains-to-build-llvm

Since it changes the requirement for building LLVM.

**Comment by linux4life798** [comment]
> I think it needs to go here: https://llvm.org/docs/ReleaseNotes.html#update-on-required-toolchains-to-build-llvm

Wow, sorry. I totally missed that section. Sounds good.

File: llvm/docs/GettingStarted.rst
Comment (linux4life798) [reply llvm/docs/GettingStarted.rst]: Done.

**Comment by linux4life798** [comment]
Corrected. Please take another look!

**Review by MaskRay — APPROVED** [review]

**Review by boomanaiden154 — APPROVED** [review]

**Comment by linux4life798** [comment]
Looks like we may have hit the first python version issue:

From the [Window x64 buildkit job](https://buildkite.com/llvm-project/github-pull-requests/builds/33515#018d492c-fee4-4595-a747-7655e3a85b41):

```
-- CTTestTidyModule ignored -- Loadable modules not supported on this platform.
-- LLD version: 19.0.0
-- Performing Test C_SUPPORTS_WERROR_IMPLICIT_FUNCTION_DECLARATION
-- Performing Test C_SUPPORTS_WERROR_IMPLICIT_FUNCTION_DECLARATION - Failed
-- Performing Test C_SUPPORTS_WERROR_MISMATCHED_TAGS
-- Performing Test C_SUPPORTS_WERROR_MISMATCHED_TAGS - Failed
-- Found Python3: C:/Python39/python.exe (found suitable version "3.9.7", minimum required is "3.8") found components: Interpreter Development Development.Module Development.Embed
CMake Error at C:/BuildTools/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Python3 (missing: Python3_NumPy_INCLUDE_DIRS NumPy) (found
  suitable version "3.9.7", minimum required is "3.8")
Call Stack (most recent call first):
  C:/BuildTools/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  C:/BuildTools/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/FindPython/Support.cmake:3165 (find_package_handle_standard_args)
  C:/BuildTools/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/FindPython3.cmake:485 (include)
  C:/ws/src/mlir/cmake/modules/MLIRDetectPythonEnv.cmake:21 (find_package)
  C:/ws/src/mlir/CMakeLists.txt:172 (mlir_configure_python_dev_packages)
-- Configuring incomplete, errors occurred!
See also "C:/ws/src/build/CMakeFiles/CMakeOutput.log".
See also "C:/ws/src/build/CMakeFiles/CMakeError.log".
```

It appears to be indicating that the version of Python is fine (3.9.7), but it couldn't find a suitable version of **NumPy**.

**Comment by boomanaiden154** [comment]
Seems like that's a MLIR Python dependency?

Numpy gets installed here in the Buildkite CI:
https://github.com/google/llvm-premerge-checks/blob/ff879c8fa1af8f06c8cb2ac7707ceef52965e87e/scripts/steps.py#L136

I wonder if there are multiple versions of Python on the windows runner and the default one isn't the 3.9.7 where the dependencies are getting installed.

**Comment by linux4life798** [comment]
Good find and good point.

**Comment by boomanaiden154** [comment]
Actually seems to be here: https://github.com/llvm/llvm-project/blob/53fea6fd6f228eb1dbd282b8d076af545dcd8228/.ci/monolithic-windows.sh#L40C64-L41C1

It seems like the configuration and the installation are using the same Python interpreter:
```
pip install -q -r C:/ws/src/mlir/python/requirements.txt
WARNING: You are using pip version 21.2.3; however, version 23.2.1 is available.
You should consider upgrading via the 'C:\Python39\python.exe -m pip install --upgrade pip' command.
```

**Comment by boomanaiden154** [comment]
> Very interesting. It looks like the same buildkite Windows x64 job now succeeds. For referencing later, it is [buildkite build #37898](https://buildkite.com/llvm-project/github-pull-requests/builds/37898). I'm going to rebase and push again to see if this buildkite build is flaky.

The Windows build (on Buildkite) is disabled currently (more context is available in this PR: https://github.com/llvm/llvm-project/pull/81538 and the discourse thread linked). It's interesting that it started passing. There have been a lot of transient failures on the Windows build that led to failures, so it wouldn't surprise me if this issue ended up being transient.

Given that the CI is disabled, I don't think there's a way to properly test this against the Buildkite Windows CI (well, you might be able to revert the patch from the disable commit in this branch), but that also shouldn't matter if it's disabled. I'm fine with this landing in its current state (especially if it passed the Windows CI before you rebased). The bigger issue either way is probably making sure everything gets past post-commit.

Some stuff within the project is also already on a newer version (https://github.com/llvm/llvm-project/issues/81598), so this landing sooner rather than later would probably be good.

Given that the RFC has consensus and the Windows CI failure is transient, I think we can probably move forward with this. Does anyone else have objections?

**Comment by linux4life798** [comment]
Much agrred.

**Comment by jyknight** [comment]
SGTM. Watch out for post submit bot failures.

**Comment by boomanaiden154** [comment]
Alright. I'm going to merge this tonight (@linux4life798 you should probably apply for commit access soonish too now that you've landed a couple commits) in the evening when buildbot traffic is hopefully low. I'll keep an eye on the post-commit bots and see what ends up failing. I'd almost expect failures, so wouldn't be surprised if it gets reverted initially, but that will let us know what breaks and we can move from there.

**Comment by boomanaiden154** [comment]
Reverted. This created a lot of post-commit failures. We'll have to get the buildbot maintainers to upgrade the Python version on those bots if we want to get this through.

https://lab.llvm.org/buildbot/#/changes/124772

**Comment by DeinAlptraum** [comment]
Is there anything happening here rn? @linux4life798 do you need any help with this?

**Comment by boomanaiden154** [comment]
Getting buildbots updated and relanding the change and potentially repeating that process is the next step. I've opened https://github.com/llvm/llvm-project/issues/83962 and emailed buildbot maintainers for bots that failed after landing this change due to old Python versions.

**Pull request merged**
